compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -605,32 +605,16 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
    *
    */
   def instantiate(tp1: NamedType, tp2: Type)(implicit ctx: Context): Type = {
-    // expose abstract type references to their bounds or tvars according to variance
-    class AbstractTypeMap(maximize: Boolean)(implicit ctx: Context) extends TypeMap {
-      def expose(lo: Type, hi: Type): Type =
-        if (variance == 0)
-          newTypeVar(TypeBounds(lo, hi))
-        else if (variance == 1)
-          if (maximize) hi else lo
-        else
-          if (maximize) lo else hi
+    // expose type param references to their bounds according to variance
+    class AbstractTypeMap(maximize: Boolean)(implicit ctx: Context) extends ApproximatingTypeMap {
+      variance = if (maximize) 1 else -1
 
       def apply(tp: Type): Type = tp match {
         case tp: TypeRef if tp.underlying.isInstanceOf[TypeBounds] =>
```

Comment:
While you're at it, maybe replace this by `tp.symbol.isTypeParam`

Reply:
@allanrenucci : `tp.symbol.isTypeParam` will miss the case where the reference is a type member.